Fix for issue #5566   intune app configuration policy

dannyKBjj opened the pull request:
#### Pull Request (PR) description
Have added a number of missing properties for the resource and added test/create support for 'Applications'

Unfortunately unable to fix 'update' due to what looks like an MgGraph issue. Added a verbose message to the code to warn the user..


reported here:
https://github.com/microsoft/Microsoft365DSC/issues/5671

and here:
 https://feedbackportal.microsoft.com/feedback/idea/c0940cc8-d7da-ef11-95f6-0022484d7a88

Similarly the module cannot support the 'Application Catalogue Settings' either. I can find them in Graph explorer but the related cmdlet doesn't work and cannot retrieve using invoke-mgGraph either.

Issue reported here
https://feedbackportal.microsoft.com/feedback/idea/f77feb49-11dc-ef11-95f6-0022484d7a88

and here
https://github.com/microsoft/Microsoft365DSC/issues/5672

#### This Pull Request (PR) fixes the following issues

- Fixes #5566


#### Task list

<!--
    To aid community reviewers in reviewing and merging your PR, please take the time to run
    through the below checklist and make sure your PR has everything updated as required.

    Change to [x] for each task in the task list that applies to your PR. For those task that
    don't apply to you PR, leave those as is.
-->

- [x ] Added an entry to the change log under the Unreleased section of the file CHANGELOG.md.
      Entry should say what was changed and how that affects users (if applicable), and
      reference the issue being resolved (if applicable).
- [ x] Resource parameter descriptions added/updated in the schema.mof.
- [x ] Resource documentation added/updated in README.md.
- [x ] Resource settings.json file contains all required permissions.
- [x ] Examples appropriately added/updated.
- [x ] Unit tests added/updated.
- [x ] New/changed code adheres to [DSC Community Style Guidelines](https://dsccommunity.org/styleguidelines).

dannyKBjj committed:
Partial fix for  issue #5566 IntuneAppConfigurationPolicy

Mostly fixes issue https://github.com/microsoft/Microsoft365DSC/issues/5566 where the applications part of the policy was completely ignored.

Module will now detect drift when Apps are changed, and is able to correctly recreate the policy when deleted. There seems to be a problem with MgGraph that means for now, 'update' will not work. Added verbose message for now, left code that should do the 'update' commented out.

Issue has been reported here: https://github.com/microsoft/Microsoft365DSC/issues/5671 and here https://feedbackportal.microsoft.com/feedback/idea/c0940cc8-d7da-ef11-95f6-0022484d7a88

dannyKBjj committed:
Partial fix for #5566 IntuneAppConfigurationPolicy module

Have added a number of missing properties for the resource and added test/create support for 'Applications'

Unfortunately unable to fix 'update' due to what looks like an MgGraph issue. Added a verbose message to the code to warn the user..

reported here:
https://github.com/microsoft/Microsoft365DSC/issues/5671

and here:
 https://feedbackportal.microsoft.com/feedback/idea/c0940cc8-d7da-ef11-95f6-0022484d7a88

Similarly the module cannot support the 'Application Catalogue Settings' either. I can find them in Graph explorer but the related cmdlet doesn't work and cannot retrieve using invoke-mgGraph either.

Issue reported here
https://feedbackportal.microsoft.com/feedback/idea/f77feb49-11dc-ef11-95f6-0022484d7a88

and here
https://github.com/microsoft/Microsoft365DSC/issues/5672

dannyKBjj commented:
Re-ran unit tests here. Work fine...

![image](https://github.com/user-attachments/assets/3a4699ab-5cb5-4e4a-b447-a8501dc9bf85)

dannyKBjj commented:
This seems to be the issue:
I need to call that command to retrieve the full App information, I don't think I can fix this at my end?!

![image](https://github.com/user-attachments/assets/d332314b-90e9-424c-bf98-8ddbcf86290f)

salbeck-sit commented:
@dannyKBjj You need to ensure that the command is defined in one of the stubs under Tests/unit/stubs. The unit-tests run by a PR won't have access to any modules

dannyKBjj committed:
Add MgBetaDeviceAppManagementTargetedManagedAppConfigurationApps to ests/unit/stubs

Add MgBetaDeviceAppManagementTargetedManagedAppConfigurationApps to ests/unit/stubs

dannyKBjj commented:
That seems to have fixed it. Thanks very much for your help!

FabienTschanz commented:
I'm taking a look at the not working update method.

FabienTschanz commented:
@dannyKBjj The following code works without issues and updates the apps on the policy: 

```powershell
if ($null -ne $Apps)
{
    $appsArray = @()
    foreach ($app in $Apps)
    {
        if ($null -ne $app.mobileAppIdentifier.bundleID)
        {
            $mobileAppIdentifierHashtable = @{
                '@odata.type' = "#microsoft.graph.iosMobileAppIdentifier"
                bundleId      = $app.mobileAppIdentifier.bundleID
            }
        }

        if ($null -ne $app.mobileAppIdentifier.packageID)
        {
            $mobileAppIdentifierHashtable = @{
                '@odata.type' = "#microsoft.graph.androidMobileAppIdentifier"
                packageId     = $app.mobileAppIdentifier.packageId
            }
        }

        if ($null -ne $app.mobileAppIdentifier.windowsAppID)
        {
            $mobileAppIdentifierHashtable = @{
                '@odata.type' = "#microsoft.graph.windowsAppIdentifier"
                windowsAppId  = $app.mobileAppIdentifier.windowsAppId
            }
        }

        $appsArray += @{
            'mobileAppIdentifier' = $mobileAppIdentifierHashtable
        }
    }

    $appsBody = @{
        appGroupType = $AppGroupType
        apps = $appsArray
    }

    Write-Verbose -Message "Updating Apps for Intune App Configuration Policy {$DisplayName}"
    $Uri = (Get-MSCloudLoginConnectionProfile -Workload MicrosoftGraph).ResourceUrl + "/beta/deviceAppManagement/targetedManagedAppConfigurations('$($currentconfigPolicy.Id)')/targetApps"
    Invoke-MgGraphRequest -Method POST -Uri $Uri -Body $($appsBody | ConvertTo-Json -Depth 10) -Verbose
}
```

I switched your `Update-MgBetaXXX` call to the `Invoke-MgGraphRequest` and slightly updated your hashtable that's used for the update. I didn't touch the create stuff, I trust that it works. 

Btw, please remove the `IsAssigned` and `DeployedAppCount` properties, they are readonly and would only make the drift incorrect as soon as you change the app count.

dannyKBjj commented:
Hi, I'd tried experimenting with invoke-mggraph myself, but had no luck. I copied you code into Set-TargetResource, and tried to use it to remediate a change I made to my 'apps' for one of my policies, but it doesn't seem to be working at my end:

![image](https://github.com/user-attachments/assets/3c697c77-9a70-4342-941c-a6a4504db8e7)

dannyKBjj commented:
[M365TenantConfig - Example.txt](https://github.com/user-attachments/files/18585487/M365TenantConfig.-.Example.txt)

I've attached an export my test policy if you want to try yourself. The policy will create just fine using the mgBeta cmdlet. If you edit the apps and try and try and update it will fail with both my code (which I commented out) and the invoke-mgGraph method you provided... or at least it is failing at my end?!

dannyKBjj commented:
Should it be PATCH not POST to update an existing policy? I tried changing the code to PATCH instead, but got the same result.

FabienTschanz commented:
> Should it be PATCH not POST to update an existing policy? I tried changing the code to PATCH instead, but got the same result.

It's a POST request to `/targetApps`. You can figure those things out if you check in the Intune portal with your browser dev tools (F12 normally) in the network tab. 

![image](https://github.com/user-attachments/assets/bc0ddb44-43f8-4618-82fb-54827638bee4)

![image](https://github.com/user-attachments/assets/5c54ce17-574e-42a2-a9f4-06dd5ea3a442)

ricmestre commented:
The app called Graph X-Ray it's also a good complement to the browser's dev tools

dannyKBjj commented:
Thanks guys, I appreciate the tips. I've been thrown in the deep-end and learning as I go along, so all help is appreciated :-D

Unfortunately the provided code just doesn't seem to be working for me?! I copy&pasted it over the if ($null -ne $Apps){} in Set-TargetResource, so don't really understand what I could have done wrong (if anything). 

It is definitely working at your end?

FabienTschanz commented:
@ricmestre Yeah that was my issue, I copy-pasted it from my test without `Get-MsCloudLoginConnectionProfile` and left the /beta, then added the full call to make it work in all clouds. It worked when I didn't prefix it with the full url. I added it because all `Invoke-MgGraphRequest` calls use the full prefix.

ricmestre commented:
This was a problem in the past and I think you actually fixed them all, or at least most of them, where the calls had graph*.microsoft.com* hardcoded, but to be honest *Get-MsCloudLoginConnectionProfile* is not even needed, but as long as it works then it shouldn't matter and it's only a preference.

![image](https://github.com/user-attachments/assets/4efb5a94-59bf-49ec-ad06-ee38bcea7f50)

dannyKBjj commented:
Hi guys, I'm sure I'm being stupid, but I experimented a fair bit with Invoke-MgGraph myself and couldn't get it to work. I also can't get the code you kindly provided to work either... I've tried everything I can think of at this point.

dannyKBjj committed:
Fixed 'Update', removed read only properties

dannyKBjj commented:
Actually, it seems to be working now. I guess I was being stupid, but I've no idea what I was doing wrong...

dannyKBjj commented:
Oh, I so now it is working!!! I guess I was being stupid. No idea what I was doing wrong before, but hey-ho.

I've pushed the changes.

FabienTschanz commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppConfigurationPolicy/MSFT_IntuneAppConfigurationPolicy.psm1:
```diff
@@ -120,19 +139,66 @@ function Get-TargetResource
+                {
+                    if($null -ne $currentValue.mobileAppIdentifier.AdditionalProperties.bundleId)
+                    {
+                        $complexMobileAppIdentifier = @{}
```
You don't need this initializer as you're specifying it just a line below.

FabienTschanz commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppConfigurationPolicy/MSFT_IntuneAppConfigurationPolicy.psm1:
```diff
@@ -120,19 +139,66 @@ function Get-TargetResource
+
+                    if($null -ne $currentValue.mobileAppIdentifier.AdditionalProperties.packageId)
+                    {
+                        $complexMobileAppIdentifier = @{}
```
You don't need this initializer as you're specifying it just a line below.

dannyKBjj replied:
removed

FabienTschanz commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppConfigurationPolicy/MSFT_IntuneAppConfigurationPolicy.psm1:
```diff
@@ -120,19 +139,66 @@ function Get-TargetResource
+
+                    if($null -ne $currentValue.mobileAppIdentifier.AdditionalProperties.windowsAppId)
+                    {
+                        $complexMobileAppIdentifier = @{}
```
You don't need this initializer as you're specifying it just a line below.

dannyKBjj replied:
removed

FabienTschanz commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppConfigurationPolicy/MSFT_IntuneAppConfigurationPolicy.psm1:
```diff
@@ -250,6 +335,41 @@ function Set-TargetResource
+            foreach($app in $Apps){
+                if($null -ne $app.mobileAppIdentifier.bundleID)
+                {
+                $mobileAppIdentifierHashtable = @{}
+                $mobileAppIdentifierHashtable['@odata.type'] = "#microsoft.graph.iosMobileAppIdentifier"
+                $mobileAppIdentifierHashtable['bundleId'] = $app.mobileAppIdentifier.bundleID
```
Indentation is not correct. Also, you could initialize the hashtable like the following: 
```powershell
$mobileAppIdentifierHashtable = @{
    '@odata.type' = "#microsoft.graph.iosMobileAppIdentifier"
    'bundleId' = $app.mobileAppIdentifier.bundleID
}
```

dannyKBjj replied:
changes made

FabienTschanz commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppConfigurationPolicy/MSFT_IntuneAppConfigurationPolicy.psm1:
```diff
@@ -120,19 +139,66 @@ function Get-TargetResource
+                    {
+                        $complexMobileAppIdentifier = @{}
+                        $complexMobileAppIdentifier = @{
+                            #'@odata.type' = "#microsoft.graph.iosMobileAppIdentifier"
```
Please remove the line if it's not needed.

dannyKBjj replied:
done

FabienTschanz commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppConfigurationPolicy/MSFT_IntuneAppConfigurationPolicy.schema.mof:
```diff
@@ -15,6 +15,21 @@ class MSFT_IntuneAppConfigurationPolicyCustomSetting
+[ClassVersion("1.0.0"), FriendlyName("MSFT_managedMobileApp")]
+class MSFT_managedMobileApp
+{
+    [Write, Description("Key of the entity.")] String id;
+    [Write, Description("Version of the entity.")] String version;
```
I'm not sure if it's clever to include the version of the mobile app since it might change and report as drifted. There is no control over it as far as I can tell in the Intune portal.

dannyKBjj replied:
Good point. Fixed.

FabienTschanz commented on CHANGELOG.md:
```diff
@@ -16,6 +16,9 @@
   * Fixes an issue with fetching a policy that does not exist.
     FIXES [#5666](https://github.com/microsoft/Microsoft365DSC/issues/5666)
+  * Adds support for testing drift in Apps setting. 
+    Set-TargetResource now able to create policies, but is still unable to update them (appears to be an issue with MgGraph).
```
This note is not relevant anymore since Update now works.

dannyKBjj replied:
ammended

dannyKBjj committed:
Requested changes implemented.

FabienTschanz commented:
Looks good so far. Some minor formatting with a space between if and the brackets could be added and the commented lines removed, but otherwise, it's ok.

dannyKBjj committed:
Requested formatting changes made.

dannyKBjj commented:
Done

FabienTschanz commented:
LGTM 🚀

dannyKBjj commented:
Hi, this got a LGTM a couple of weeks back but hasn't been merged?

FabienTschanz commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppConfigurationPolicy/MSFT_IntuneAppConfigurationPolicy.psm1:
```diff
@@ -561,6 +771,16 @@ function Export-TargetResource
                 $currentDSCBlock = Convert-DSCStringParamToVariable -DSCBlock $currentDSCBlock -ParameterName 'Assignments' -IsCIMArray:$isCIMArray
             }
 
+            if ($Results.Apps)
+            {
+                $isCIMArray = $false
+                if ($Results.Apps.getType().Fullname -like '*[[\]]')
+                {
+                    $isCIMArray = $true
+                }
+                $currentDSCBlock = Convert-DSCStringParamToVariable -DSCBlock $currentDSCBlock -ParameterName 'Apps' -IsCIMArray:$isCIMArray
+            }
+
```
One thing that was changed in the latest PRs was that `Convert-DSCStringParamToVariable` is no longer necessary, as well as `$Results = Update-M365DSCExportAuthenticationResults -ConnectionMode $ConnectionMode -Results $Results`. You can safely remove the lines 759 (the first if check for `CustomSettings`) all the way to 782 (last bracket of the if check for `Apps`). 

Attach the following code to `-Credential $Credential`: 

```powershell
...
-Credential $Credential `
-NoEscape @('CustomSettings', 'Assignments', 'Apps')
```

dannyKBjj replied:
made requested changes.

dannyKBjj committed:
requested change made

dannyKBjj committed:
Merge branch 'Dev' into Fix-for-issue-#5566---IntuneAppConfigurationPolicy

dannyKBjj commented:
Hi, I made requested changes. I tested with the latest version of m365DSC and work was broken.

[2025/02/17 01:57:53]
{ObjectNotFound}
System.Management.Automation.CommandNotFoundException: The term '-NoEscape' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
   at System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext funcContext, Exception exception)
   
I decided to upload anyway, as perhaps this is some sort of unreleased change, but now my unit tests are failing on Export... so not sure what to do?

dannyKBjj committed:
Backed out requested changes that seemed to break module

dannyKBjj committed:
Merge branch 'Fix-for-issue-#5566---IntuneAppConfigurationPolicy' of https://github.com/dannyKBjj/Microsoft365DSC into Fix-for-issue-#5566---IntuneAppConfigurationPolicy

dannyKBjj committed:
pulled origin recommitting

dannyKBjj committed:
backed out requested changes as they appear to break things.

dannyKBjj commented:
Hi, I made requested changes. I tested with the latest version of m365DSC and module was broken.

[2025/02/17 01:57:53]
{ObjectNotFound}
System.Management.Automation.CommandNotFoundException: The term '-NoEscape' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
   at System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext funcContext, Exception exception)
   
I decided to upload anyway, as perhaps this is some sort of unreleased change, but then my unit tests start failing for Export. Thinking the changes are what caused the problem, I backed them out, but the unit tests are still failing on the export/reverseDSC tests.

I'm at a bit of a loss here...

dannyKBjj committed:
Merge remote-tracking branch 'upstream/Dev' into Fix-for-issue-#5566---IntuneAppConfigurationPolicy

FabienTschanz commented:
@dannyKBjj The `-NoEscape` parameter was recently added. Update your dependency for ReverseDSC to the latest version 2.0.0.27 and the tests should run fine.

FabienTschanz commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppConfigurationPolicy/MSFT_IntuneAppConfigurationPolicy.psm1:
```diff
@@ -376,6 +558,12 @@ function Test-TargetResource
+    {
+        Write-Verbose -Message "Test-TargetResource returned $false"
+        return $false
+    }
```
This is not necessary anymore, the property `Ensure` will be tested further below with `Test-M365DSCParameterState`.

dannyKBjj replied:
done

FabienTschanz commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppConfigurationPolicy/MSFT_IntuneAppConfigurationPolicy.psm1:
```diff
@@ -533,13 +749,38 @@ function Export-TargetResource
 
-
+            $Results = Update-M365DSCExportAuthenticationResults -ConnectionMode $ConnectionMode `
+                -Results $Results
```
Please remove this call. Not necessary anymore.

dannyKBjj replied:
removed

FabienTschanz commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppConfigurationPolicy/MSFT_IntuneAppConfigurationPolicy.psm1:
```diff
@@ -533,13 +749,38 @@ function Export-TargetResource
+                    $isCIMArray = $true
+                }
+                $currentDSCBlock = Convert-DSCStringParamToVariable -DSCBlock $currentDSCBlock -ParameterName 'Apps' -IsCIMArray:$isCIMArray
+            }
```
```suggestion
                -Credential $Credential `
                -NoEscape @('Apps', 'CustomSettings', 'Assignments')
```

This was just recently changed. Please update it to my suggestion. `Convert-DSCStringParamToVariable` is not necessary anymore for CIM instances to be called separately.

dannyKBjj replied:
done

dannyKBjj committed:
Updated Export-DSCResource

dannyKBjj commented:
Unit tests passed, hopefully ready to go!

NikCharlebois reviewed (changes requested):
Please update your branch with the latest. Changelog is outdated.

dannyKBjj committed:
Merge remote-tracking branch 'upstream/Dev' into Fix-for-issue-#5566---IntuneAppConfigurationPolicy

dannyKBjj committed:
Updated from Dev Updated changelog.md

Requested changes

dannyKBjj commented:
Updated and corrected change log

dannyKBjj committed:
Merge branch 'Dev' into Fix-for-issue-#5566---IntuneAppConfigurationPolicy

dannyKBjj commented:
any chance we can merge this soonish before it gets horribly out of date again, got its first LGTM 3 weeks ago now? :-)

ykuijs commented:
@dannyKBjj, sorry for the delay. I have been a little swamped lately. Doing a final review now.

ykuijs committed:
Update CHANGELOG.md with new IntuneAppConfigurationPolicy entries

ykuijs committed:
Merge branch 'Dev' into Fix-for-issue-#5566---IntuneAppConfigurationPolicy
